src/codegen/baseline_jit.h

```diff
@@ -0,0 +1,249 @@
+//
+// To execute a specific CFGBlock one has to call:
+//      CFGBlock* block;
+//      block->entry_code(ast_interpreter_instance, block)
+//
+// Signature of a JitCodeBlock:
+//  std::pair<CFGBlock*, Box*>(*entry_code)(ASTInterpreter* interp, CFGBlock* block)
+//  args:
+//      interp:  instance to the ASTInterpreter
+//      block:   block to execute
+//
+//  return value:
+//      first:   next block to execute in the interpreter
+//                  if 0, we executed a return node and should return second
+//      second:  return value only used if first == 0
+//
+// Basic layout of generated code block is:
+// entry_code:
+//      push   %rbp                 ; setup frame pointer
+//      mov    %rsp,%rbp
+//      sub    $0x108,%rsp          ; setup scratch, 0x108 = scratch_size + 8 (=stack alignment)
+//      push   %rdi                 ; save the pointer to ASTInterpreter instance
+//      jmpq   *0x8(%rsi)           ; jump to block->code
+//                                      possible values: first_JitFragment, second_JitFragment,...
+//
+// first_JitFragment:
+//      ...
+//    ; Side exit (e.g. if <val> <block1> else <block2>
+//      movabs $0x1270014108,%rcx   ; rcx = True
+//      cmp    %rax,%rcx            ; rax == True
+//      jne    end_side_exit
+//      mov    %rax,0x10(%rsp)      ;
+//      movabs $0x215bb60,%rax      ; rax = CFGBlock* to call next (rax is also the 1. return reg)
+//      mov    0x8(%rax),%rsi       ; load CFGBlock->code
+//      test   %rsi,%rsi            ; CFGBlock->code == 0
+//      je     epilog               ; exit to interpreter if code == 0
+//      jmpq   *0x8(%rax)           ; jump to new jit fragment (e.g second_JitFragment)
+//    end_side_exit:
+//      ....
+
+// second_JitFragment:
+//      ...
+//
+// nth_JitFragment:
+//      ...                         ; direct jump previous JITed block
+//      jmp first_JitFragment
+//
+// epilog:                          ; code which jumps to epilog has to make sure that
+//                                  ; rax contains the next block to execute
+//                                  : or 0 if we are finished but then rdx must contain the Box* value to return
+//      leave
+//      ret
+//
+class JitCodeBlock {
+private:
+    static constexpr int scratch_size = 256;
+    static constexpr int code_size = 4096 * 2;
+    static constexpr int epilog_size = 2; // size of [leave, ret] in bytes
+
+    void* code;
+    int entry_offset;
+    int epilog_offset;
+    assembler::Assembler a;
+    bool iscurrently_tracing;
+    bool asm_failed;
+
+public:
+    JitCodeBlock(llvm::StringRef name);
+
+    std::unique_ptr<JitFragment> newFragment(CFGBlock* block, int jump_offset = 0);
+    bool shouldCreateNewBlock() const { return asm_failed || a.bytesLeft() < 128; }
+    void fragmentFinished(int size, bool not_enough_space);
+};
+
+class JitFragment : public Rewriter {
```

Comment:
Maybe we could call it "JitFragmentWriter"?

I'm not super psyched that this is a subclass of Rewriter -- that doesn't feel quite right to me.  I might have said that it has-a rewriter, but it looks like we use some of the internal APIs so that might not be easy.  I think there might be some fundamental differences in what they need to do, so I guess has-a isn't really right either.  Maybe we should (in the future) refactor out the common aspects of them (register allocation, action registration, etc) and then make them both subclasses of some common class?